types.go

```diff
@@ -1819,38 +1857,41 @@ type RestrictedType struct {
 }
 
 func NewRestrictedType(
-	typeID string,
 	typ Type,
 	restrictions []Type,
 ) *RestrictedType {
 	return &RestrictedType{
-		typeID:       typeID,
 		Type:         typ,
 		Restrictions: restrictions,
 	}
 }
 
 func NewMeteredRestrictedType(
 	gauge common.MemoryGauge,
-	typeID string,
 	typ Type,
 	restrictions []Type,
 ) *RestrictedType {
 	common.UseMemory(gauge, common.CadenceRestrictedTypeMemoryUsage)
-	return NewRestrictedType(typeID, typ, restrictions)
+	return NewRestrictedType(typ, restrictions)
 }
 
 func (*RestrictedType) isType() {}
 
 func (t *RestrictedType) ID() string {
+	if t.typeID == "" {
+		var restrictionStrings []string
+		restrictionCount := len(t.Restrictions)
+		if restrictionCount > 0 {
+			restrictionStrings = make([]string, 0, restrictionCount)
+			for _, restriction := range t.Restrictions {
```

Comment:
would it be better to use the `t.restrictionSet` instead of `t.Restrictions`?

Reply:
No, `restrictionSet` is a `map`, so the order of iteration is non-deterministic.

Unrelated, but optimized the restriction set in 3f90c885e20d9b022764a930824c2511aa96e4ef